README.md
# PSInsight
- - - - 
PSInsight is a Powershell wrapper to interface with Jira Insight's [API](https://documentation.mindville.com/display/INSCLOUD/REST+API).

PSInsight can be used to build or modify the [Jira Insight](https://marketplace.atlassian.com/apps/1212137/insight-asset-management?hosting=cloud&tab=overview) schema or to add or modify assets within the asset management database.

Links to Insight API information for each function can be found under links via Get-Help

```
Get-Help Get-InsightObjectSchema
```

# Getting Started #
## Using PowershellGallery ##
```
Install-Module PSInsight
Import-Module PSInsight
```

## Using Git ##
Clone the repository.
```
git clone "https://github.com/DamagedDingo/PSInsight.git"
```
Place directory into a module directory (e.g. $env:USERPROFILE\Documents\WindowsPowerShell\Modules).
```
Move-Item -path ".\psinsight\psinsight" -Destination "$env:USERPROFILE\Documents\WindowsPowerShell\Modules"
```
Import the module.
```
Import-Module PSInsight
```

# Using your API Key #
Almost all commands require an API key.
  
For ease of use, each command looks for the variable automatically in the following order:  
    In the global scope for InsightApiKey  
    As passed as parameters to the command  
    As an AutomationVariable  
    A prompt to host to enter Key manually  

# Example Script #
The Examples folder contains samples on how to build the schema as well as object types.
The corrisponding JSON file will show how to structure the data for the builder scripts.

![Schema](https://i.imgur.com/NByWKhl.png)
![Attributes](https://i.imgur.com/GzvExXL.png)

# Available Functions #
Use get-help for more information about each function.

## Icons ##
```
Get-InsightIcons  
```
## Object Schema ##
```
Get-InsightObjectSchema
New-InsightObjectSchema
Remove-InsightObjectSchema
Set-InsightObjectSchema
```
## Object Type Attributes ##
```
Get-InsightObjectTypeAttributes
New-InsightObjectTypeAttributes
Remove-InsightObjectTypeAttributes
Set-InsightObjectTypeAttributes
```
## Object Types ##
```
Get-InsightObjectTypes
New-InsightObjectTypes
Remove-InsightObjectTypes
Set-InsightObjectTypes
```
## Objects ##
```
Get-InsightObject
New-InsightObject
New-InsightObjectAttribute
Remove-InsightObject
Set-InsightObject
```
## Statuses ##
```
Get-InsightStatuses
New-InsightStatuses
Remove-InsightStatuses
Set-InsightStatuses
```
## Build ##
These are custom functions to build out your Schema and ObjectTypes as per the example scripts.
```
Build-Schema
Build-ObjectTypes
Test-Module
```
## Private ##
```
New-InsightHeaders
```

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

Add comment under [Unreleased] and commit with !Deploy!
Format as below (Requires the '### Added' as a header or task fails)
...### Added
...- Move Test-Module from Private to public\build
...- Update ReadMe.MD

## [Unreleased]

## [1.25.20210511.0] - 2021-05-11
### Added
- Missing attribute 'uniqueAttribute' in New-InsightObjectTypeAttributes and Set-InsightObjectTypeAttributes

## [1.24.20210511.0] - 2021-05-11
### Added
- fix 'UniqueAttribute' in New-InsightObjectTypeAttributes and Set-InsightObjectTypeAttributes

## [1.23.20210511.0] - 2021-05-11
### Added
- Added 'typeValueMulti' to status in New-InsightObjectTypeAttributes

## [1.22.20210507.0] - 2021-05-07
### Added
- Convert-ChassisType ID to accept [string]
- Updated Get-MEMCMSQL_ExtendedInfo to accept [string] - SCCM stores as string not [int]

## [1.21.20210507.0] - 2021-05-07
### Added
- Convert-ChassisType ID to accept [int]
- Updated Get-MEMCMSQL_ExtendedInfo


## [1.20.20210506.0] - 2021-05-06
### Added
- Update Test-Module
- Add Join-Object

## [1.19.20210430.0] - 2021-04-30
### Added
- Update cmdlet name Get-ZoomRoomsDashboard

## [1.18.20210428.0] - 2021-04-28
### Added
- Remove Schema name from 'Get-ZoomRoomsObjects'
- Update 'Set-InsightObject'

## [1.17.20210422.0] - 2021-04-22
### Added
- Updated Set-Insight Object to add -ShowJSON function

## [1.16.20210421.0] - 2021-04-21
### Added
- Find-InsightObjectsAdvanced bug fix for attributesToDisplay (Still seems to be bug at Insight end)
- Find-InsightObjectsAdvanced Added ShowJSON switch that will show json and break out of function. Good for troubleshooting
- ObjectTypes.json updated to include status field.
- Connect-MEMCM cleaned up and removed unwanted code
- New-InsightObjectTypeAttributes updated JSON depth.. descriptions still not working logged ticket to Insight
- New-InsightObjectAttribute Removed whitespace
- Build-ObjectType Removed whitespace and test code.

## [1.15.20210414.0] - 2021-04-14
### Added
- Build-ObjectType Add description parameter
- New-InsightObjectTypeAttributes Add description parameter

## [1.14.20210413.0] - 2021-04-13
### Added
- Fixed bug in zoom room builder where parent_room was not created or populated.
- Added field 'typeValue' for type 'Object' in New-InsightObjectTypeAttributes
- Added Find-InsightObjects
- Added Find-InsightObjectsAdvanced
- Removed Links from JSON and Build-ObjectType

## [1.13.20210412.0] - 2021-04-12
### Added
- Update Convert-ChassisType to contain catagory as well as ID.
- Update ObjectTypes.json - Remove Hostname, Update SerialName field, add links
- Connect-MEMCM updated
- Update Get-MEMCMSQL_ExtendedInfo - New Chasis Info, SerialName
- Statuses examples and builder

## [1.12.20210409.0] - 2021-04-09
### Added
- Move -raw on get-content to stop positional errors
- Added 'Get-MEMCMSQL_ExtendedInfo' to collect extra properties for computers pulled from SCCM
- Added SiteServer.Json example
- Updated Zoom Rooms Example.ps1
- Updated Example_ObjectTypeBuilder.ps1
- Updated Example_Schemabuilder.ps1
- Added Convert-ChassisType.ps1
- Added Get-ZoomRoomsObjects.ps1


## [1.11.20210408.0] - 2021-04-08
### Added
- Rearranged ObjectTypes.json
- Added 'SCCM Example.ps1' (Currently only building Computer Object..)

## [1.10.20210408.0] - 2021-04-08
### Added
- 'Test-Module PSInsight' added to examples to ensure module is loaded
- Populate_ZoomRooms sample script added

## [1.9.20210408.0] - 2021-04-08
### Added
- Renamed Build Folder to Builder as 'Build' is a reserved folder for github actions. 
- Edit ReadMe.MD 

## [1.8.20210408.0] - 2021-04-08
### Added
- Move Test-Module from Private to public\build
- Update ReadMe.MD

## [1.7.20210408.0] - 2021-04-08
### Added
- Updated ReadMe.md
- Build-ObjectType and Build-Schema added to build the database from JSON input (See examples folder).
- Examples\Schema added to provide sample data and script to automaticly build a Schema.
- Examples\ObjectTypes added to provide sample data and script to automaticly build ObjectTypes within the schema.
- Superseeded example files removed


## [1.6.20210407.0] - 2021-04-07
### Added
- Update New-InsightObjectTypes to support more parameters (abstractObjectType & inherited).

## [1.5.20201023.0] - 2020-10-23
### Added
- Updated the format of the dynamic parameters based on the help from chrisdent (https://discord.gg/winadmin ) makes it much more readable and can remove all teh commenting. 

## [1.4.20201022.0] - 2020-10-22
### Added
- Fixed issue with APIkey

## [1.3.20201021.0] - 2020-10-21
### Added
- Added Recurse to the Invoke-Build Script for the PSM1

## [1.2.20201021.0] - 2020-10-21
### Added
- Added Recurse to the Invoke-Build Script

## [1.1.20201021.0] - 2020-10-21
### Added
- First Deployment

## [1.0.0] - 2020-10-15
### Added
- Added to start versioning[Unreleased]: https://github.com/DamagedDingo/PSInsight/compare/1.25.20210511.0..HEAD
[1.25.20210511.0]: https://github.com/DamagedDingo/PSInsight/compare/1.24.20210511.0..1.25.20210511.0
[1.24.20210511.0]: https://github.com/DamagedDingo/PSInsight/compare/1.23.20210511.0..1.24.20210511.0
[1.23.20210511.0]: https://github.com/DamagedDingo/PSInsight/compare/1.22.20210507.0..1.23.20210511.0
[1.22.20210507.0]: https://github.com/DamagedDingo/PSInsight/compare/1.21.20210507.0..1.22.20210507.0
[1.21.20210507.0]: https://github.com/DamagedDingo/PSInsight/compare/1.20.20210506.0..1.21.20210507.0
[1.20.20210506.0]: https://github.com/DamagedDingo/PSInsight/compare/1.19.20210430.0..1.20.20210506.0
[1.19.20210430.0]: https://github.com/DamagedDingo/PSInsight/compare/1.18.20210428.0..1.19.20210430.0
[1.18.20210428.0]: https://github.com/DamagedDingo/PSInsight/compare/1.17.20210422.0..1.18.20210428.0
[1.17.20210422.0]: https://github.com/DamagedDingo/PSInsight/compare/1.16.20210421.0..1.17.20210422.0

Examples/Objects/SCCM/SCCM Example.ps1
#Required Modules
Test-Module "PSInsight"
Test-Module "SQLServer"
Test-Module "Join-Object"

$CollectionName = "All Desktops"

# Site configuration and database
$Server = "" # SMS Provider machine name
$SCCMDB = "CM_001" # Database name
$SiteCode = "001" # Site code 

if(($host.Name -match 'consolehost')) {$true}

# Connect to SCCM
$initParams = @{}
if((Get-Module ConfigurationManager) -eq $null) {
    Import-Module "$($ENV:SMS_ADMIN_UI_PATH)\..\ConfigurationManager.psd1" @initParams 
}
if((Get-PSDrive -Name $SiteCode -PSProvider CMSite -ErrorAction SilentlyContinue) -eq $null) {
    New-PSDrive -Name $SiteCode -PSProvider CMSite -Root $Server @initParams
}
Set-Location "$($SiteCode):\" @initParams

# Get list of computers
$Desktops = Get-CMCollectionMember -CollectionName $CollectionName | Select Name,ResourceID,CoManaged,DeviceOS,DeviceOSBuild,Domain,IsVirtualMachine,LastHardwareScan,LastLogonUser,MACAddress,PrimaryUser

# Get computer information from SCCM SQL
$v_GS_COMPUTER_SYSTEM = Invoke-Sqlcmd -ServerInstance $Server -Database $SCCMDB -Query "Select * From v_GS_COMPUTER_SYSTEM"
$v_GS_PHYSICAL_MEMORY = Invoke-Sqlcmd -ServerInstance $Server -Database $SCCMDB -Query "Select * From v_GS_PHYSICAL_MEMORY"
$v_GS_MAPPEDDRIVES640 = Invoke-Sqlcmd -ServerInstance $Server -Database $SCCMDB -Query "Select * From v_GS_MAPPEDDRIVES640"
$v_GS_DISK = Invoke-Sqlcmd -ServerInstance $Server -Database $SCCMDB -Query "Select * From v_GS_DISK"
$v_GS_OPERATING_SYSTEM = Invoke-Sqlcmd -ServerInstance $Server -Database $SCCMDB -Query "Select * From v_GS_OPERATING_SYSTEM"
$v_GS_LOGICAL_DISK = Invoke-Sqlcmd -ServerInstance $Server -Database $SCCMDB -Query "Select ResourceID,Size0,FreeSpace0 From v_GS_LOGICAL_DISK"
$v_GS_SYSTEM_ENCLOSURE = Invoke-Sqlcmd -ServerInstance $Server -Database $SCCMDB -Query "Select ResourceID,ChassisTypes0 From v_GS_SYSTEM_ENCLOSURE"
$v_GS_X86_PC_MEMORY = Invoke-Sqlcmd -ServerInstance $Server -Database $SCCMDB -Query "Select ResourceID,TotalPhysicalMemory0 From v_GS_X86_PC_MEMORY"
$v_GS_VIRTUAL_MACHINE = Invoke-Sqlcmd -ServerInstance $Server -Database $SCCMDB -Query "Select * From v_GS_VIRTUAL_MACHINE"

if ([Environment]::UserInteractive -eq $true) {
    $DesktopCount = $Desktops.count
    $i = 0
}


$Assets = ForEach ($Desktop in $Desktops) {

    if ($DesktopCount) {
        $Percentage = [math]::Round($(($i / $DesktopCount)*100),2)

        $Progress = @{
            Activity = "Building Computer Objects"
            Status = "$Percentage% Complete:"
            PercentComplete = $Percentage
        }
        # add SecondsRemaining
        Write-Progress @Progress
        $i++
    }
    

    $desktop | Add-Member -MemberType NoteProperty -Name 'v_GS_COMPUTER_SYSTEM' -Value $($v_GS_COMPUTER_SYSTEM | Where { $_.ResourceID -like $Desktop.ResourceID }) -Force
    $desktop | Add-Member -MemberType NoteProperty -Name 'v_GS_PHYSICAL_MEMORY' -Value $($v_GS_PHYSICAL_MEMORY | Where { $_.ResourceID -like $Desktop.ResourceID }) -Force
    $desktop | Add-Member -MemberType NoteProperty -Name 'v_GS_OPERATING_SYSTEM' -Value $($v_GS_OPERATING_SYSTEM | Where { $_.ResourceID -like $Desktop.ResourceID }) -Force
    $desktop | Add-Member -MemberType NoteProperty -Name 'v_GS_MAPPEDDRIVES640' -Value $($v_GS_MAPPEDDRIVES640 | Where { $_.ResourceID -like $Desktop.ResourceID }) -Force
    $desktop | Add-Member -MemberType NoteProperty -Name 'v_GS_DISK' -Value $($v_GS_DISK | Where { $_.ResourceID -like $Desktop.ResourceID }) -Force

    $desktop | Add-Member -MemberType NoteProperty -Name 'v_GS_LOGICAL_DISK' -Value $($v_GS_LOGICAL_DISK | Where { $_.ResourceID -like $Desktop.ResourceID }) -Force
    $desktop | Add-Member -MemberType NoteProperty -Name 'v_GS_SYSTEM_ENCLOSURE' -Value $($v_GS_SYSTEM_ENCLOSURE | Where { $_.ResourceID -like $Desktop.ResourceID }) -Force
    $desktop | Add-Member -MemberType NoteProperty -Name 'v_GS_X86_PC_MEMORY' -Value $($v_GS_X86_PC_MEMORY | Where { $_.ResourceID -like $Desktop.ResourceID }) -Force


    switch ($Desktop.v_GS_SYSTEM_ENCLOSURE.ChassisTypes0) {

        1 { $ChassisName = "Other" }
        2 { $ChassisName = "Unknown" }
        3 { $ChassisName = "Desktop" }
        4 { $ChassisName = "Low Profile Desktop" }
        5 { $ChassisName = "Pizza Box" }
        6 { $ChassisName = "Mini Tower" }
        7 { $ChassisName = "Tower" }
        8 { $ChassisName = "Portable" }
        9 { $ChassisName = "Laptop" }
        10 { $ChassisName = "Notebook" }
        11 { $ChassisName = "Hand Held" }
        12 { $ChassisName = "Docking Station" }
        13 { $ChassisName = "All in One" }
        14 { $ChassisName = "Sub Notebook" }
        15 { $ChassisName = "Space-Saving" }
        16 { $ChassisName = "Lunch Box" }
        17 { $ChassisName = "Main System Chassis" }
        18 { $ChassisName = "Expansion Chassis" }
        19 { $ChassisName = "SubChassis" }
        20 { $ChassisName = "Bus Expansion Chassis" }
        21 { $ChassisName = "Peripheral Chassis" }
        22 { $ChassisName = "Storage Chassis" }
        23 { $ChassisName = "Rack Mount Chassis" }
        24 { $ChassisName = "Sealed-Case PC" }
        30 { $ChassisName = "Tablet" }
        31 { $ChassisName = "Convertible" }
        32 { $ChassisName = "Detachable" }
        Default { $ChassisName = "" }
    }

    $Asset = [PSCustomObject]@{
        "Hostname" = $desktop.Name
        "ResourceID" = $desktop.ResourceID
        "Domain" = $desktop.Domain
        "IsVirtualMachine" = $desktop.IsVirtualMachine
        "LastHardwareScan" = $desktop.LastHardwareScan
        "LastLogonUser" = $desktop.LastLogonUser
        "PrimaryUser" = $desktop.PrimaryUser
        "MACAddress" = $desktop.MACAddress
        
        # v_GS_COMPUTER_SYSTEM
        "Manufacturer" = $Desktop.v_GS_COMPUTER_SYSTEM.Manufacturer0
        "Model" = $Desktop.v_GS_COMPUTER_SYSTEM.Model0
        "TotalPhysicalMemory" = $Desktop.v_GS_COMPUTER_SYSTEM.TotalPhysicalMemory0

        # v_GS_SYSTEM_ENCLOSURE
        "ChassisTypes" = $ChassisName
        "ChassisTypesCode" = $Desktop.v_GS_SYSTEM_ENCLOSURE.ChassisTypes0

        # v_GS_OPERATING_SYSTEM  
        "Caption" = $Desktop.v_GS_OPERATING_SYSTEM.Caption0
        "CSDVersion" = $Desktop.v_GS_OPERATING_SYSTEM.CSDVersion0
        "InstallDate" = $Desktop.v_GS_OPERATING_SYSTEM.InstallDate0
        "OSArchitecture" = $Desktop.v_GS_OPERATING_SYSTEM.OSArchitecture0
        "Version" = $Desktop.v_GS_OPERATING_SYSTEM.Version0

        # v_GS_PHYSICAL_MEMORY - Physical Memory (Desktop?)
        "DeviceLocator" = $Desktop.v_GS_PHYSICAL_MEMORY.DeviceLocator0
        "FormFactor" = $Desktop.v_GS_PHYSICAL_MEMORY.FormFactor0
        "Capacity" = $Desktop.v_GS_PHYSICAL_MEMORY.Capacity0
        "Speed" = $Desktop.v_GS_PHYSICAL_MEMORY.Speed0

        # $v_GS_X86_PC_MEMORY
        "TotalPhysicalMemory0" = $Desktop.v_GS_X86_PC_MEMORY.TotalPhysicalMemory0

        # v_GS_DISK - Disk (Desktop?)
        "Disk_Caption" = $desktop.v_GS_DISK.Caption0
        "Disk_DeviceID" = $desktop.v_GS_DISK.DeviceID0
        "Disk_InterfaceType" = $desktop.v_GS_DISK.InterfaceType0
        "Disk_Manufacturer" = $desktop.v_GS_DISK.Manufacturer0
        "Disk_MediaType" = $desktop.v_GS_DISK.MediaType0
        "Disk_Model" = $desktop.v_GS_DISK.Model0

        # v_GS_LOGICAL_DISK
        "Disk_Size" = $desktop.v_GS_LOGICAL_DISK.Size0
        "Disk_FreeSpace" = $desktop.v_GS_LOGICAL_DISK.FreeSpace0

        # v_GS_MAPPEDDRIVES640ISK - Mapped Drives
        "DriveLetter" = $desktop.v_GS_MAPPEDDRIVES640.DriveLetter0
        "ShareName" = $desktop.v_GS_MAPPEDDRIVES640.ShareName0
        "UserDomain" = $desktop.v_GS_MAPPEDDRIVES640.UserDomain0
        "UserName" = $desktop.v_GS_MAPPEDDRIVES640.UserName0

    }
    $Asset
    }

    